pdg/src/builder.rs

```diff
@@ -102,42 +104,118 @@ impl EventKindExt for EventKind {
+    mut mapping: ProvenanceInfo,
 ) {
     use EventKind::*;
     match *event_kind {
-        Alloc { ptr, .. } => {
+        Alloc { ptr, size, .. } => {
+            // TODO: check for overlap with existing allocations
+            mapping.size = Some(size);
             provenances.insert(ptr, mapping);
         }
-        CopyPtr(ptr) => {
-            // only insert if not already there
-            if let Err(..) = provenances.try_insert(ptr, mapping) {
-                log::warn!("0x{:x} already has a source", ptr);
-            }
-        }
-        Realloc { new_ptr, .. } => {
+        Realloc { new_ptr, size, .. } => {
+            mapping.size = Some(size);
             provenances.insert(new_ptr, mapping);
         }
-        Offset(_, _, new_ptr) => {
-            provenances.insert(new_ptr, mapping);
+        Free { ptr } if ptr != 0 => {
+            // TODO: handle the case where `ptr` falls inside an allocation
+            let old_prov = provenances.remove(&ptr);
+            if old_prov.is_none() {
+                log::warn!("Tried to free invalid pointer 0x{:x}", ptr);
+            }
         }
-        CopyRef => {
-            provenances.insert(metadata.destination.clone().unwrap().local.into(), mapping);
+        CopyPtr(ptr) | Offset(_, _, ptr) | Project(_, ptr, _) => {
+            // Check that the pointer falls inside an existing allocation
+            let need_insert = provenances
+                .range(0..=ptr)
+                .next_back()
+                .map(|(start, pi)| {
+                    // If the existing pointer has a size, check it
+                    // Otherwise, insert ours if it's different
+                    let size = match pi.size {
+                        Some(size) => size,
+                        None => return ptr != *start,
+                    };
+
+                    let end = start.saturating_add(size);
+                    if (*start..=end).contains(&ptr) {
+                        return false;
+                    }
+
+                    if matches!(*event_kind, CopyPtr(_)) {
+                        // A new pointer that we don't know where it came from
+                        // Insert it for now, we should decide later what to do
+                        //
+                        // Assume an unknown pointer points to a word-sized object
+                        log::warn!("Pointer of unknown origin 0x{:x}", ptr);
+                        return true;
+                    }
+
+                    log::warn!(
+                        "0x{:x} outside of object bounds 0x{:x}-0x{:x}",
+                        ptr,
+                        start,
+                        end
+                    );
+
+                    false
+                })
+                .unwrap_or(true);
+
+            if need_insert {
+                provenances.insert(ptr, mapping);
+            }
         }
-        AddrOfLocal(ptr, _) => {
+        CopyRef => unimplemented!(),
```

Comment:
Is it okay for this to remain unimplemented?  Could this cause trouble when running on partially refactored code that uses safe refs in some places?

Reply:
The actual problem is the dynamic instrumentation doesn't ever emit `CopyRef` right now. I'm wondering if we should just remove this node kind.

Reply:
Update: I removed `CopyRef`.